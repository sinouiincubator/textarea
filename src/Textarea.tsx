import React, { useState, useCallback } from 'react';
import styled, { css } from 'styled-components';
import classNames from 'classnames';

export interface TextareaProps {
  /**
   * 文本框的值
   */
  value?: string;
  /**
   * 文本框的默认值
   */
  defaultValue?: string;
  /**
   * 值变更时的回调函数
   */
  onChange?: (event: React.ChangeEvent<HTMLTextAreaElement>) => void;
  /**
   * 失去焦点时的回调函数
   */
  onBlur?: () => void;
  /**
   * 是都自动获取焦点
   */
  autoFocus?: boolean;
  /**
   * 最大长度
   */
  maxLength?: number;
  /**
   * 提示文本
   */
  placeholder?: string;
  /**
   * 样式
   */
  style?: React.CSSProperties;
  /**
   * 自定义样式类名称
   */
  className?: string;
}

const borderStyle = css`
  border: 1px solid ${(props) => props.theme.palette.input.bottomLine};
  transition: border-color 300ms;

  :hover {
    padding: 3px;
    border: 2px solid ${(props) => props.theme.palette.action.active};
  }
`;

const focusedBorderStyle = css`
  padding: 3px;
  border: 2px solid ${(props) => props.theme.palette.primary[500]};
`;

const StyledTextarea = styled.textarea<TextareaProps & { focused?: boolean }>`
  min-height: 80px;
  box-sizing: border-box;
  padding: 4px;
  resize: none;
  outline: none;
  font-size: ${(props) => props.theme.typography.subheading.fontSize}rem;
  font-family: ${(props) => props.theme.typography.fontFamily};
  font-weight: ${(props) => props.theme.typography.fontWeightRegular};
  overflow: auto;
  color: ${(props) => props.theme.palette.text.primary};
  background-color: transparent;
  ${(props) => (props.focused ? focusedBorderStyle : borderStyle)};
`;

/**
 * 多行文本输入框
 * @param props TextareaProps
 */
export default function Textarea(props: TextareaProps) {
  const { defaultValue, value, onChange, onBlur, className } = props;
  const [focused, setFocused] = useState(false);

  /**
   * 获取焦点时的回调函数
   */
  const handleOnFocus = useCallback(() => setFocused(true), []);

  /**
   * 失去焦点时的回调函数
   */
  const handleBlur = useCallback(() => {
    setFocused(false);
    if (onBlur) {
      onBlur();
    }
  }, [onBlur]);

  return (
    <StyledTextarea
      {...props}
      defaultValue={defaultValue}
      value={value}
      onChange={onChange}
      onFocus={handleOnFocus}
      focused={focused}
      onBlur={handleBlur}
      className={classNames('sinouiincubator-textarea', className)}
    />
  );
}
