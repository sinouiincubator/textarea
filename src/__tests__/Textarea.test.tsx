import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import { render, fireEvent, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@sinoui/theme';
import Textarea from 'src/Textarea';

describe('Textarea 镜像测试', () => {
  it('渲染Textarea', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <Textarea />
        </ThemeProvider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('onchange', () => {
    const onChage = jest.fn();
    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <Textarea value="001" onChange={onChage} />
        </ThemeProvider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe('验收测试', () => {
  it('渲染', () => {
    const { getByPlaceholderText } = render(
      <ThemeProvider theme={defaultTheme}>
        <Textarea placeholder="请输入内容" />
      </ThemeProvider>,
    );

    expect(getByPlaceholderText('请输入内容')).toBeDefined();
  });

  it('值变更', () => {
    const onChange = jest.fn();
    const { getByPlaceholderText } = render(
      <ThemeProvider theme={defaultTheme}>
        <Textarea placeholder="请输入内容" onChange={onChange} />
      </ThemeProvider>,
    );

    fireEvent.change(getByPlaceholderText('请输入内容'), {
      target: { value: '001' },
    });

    expect(onChange).toBeCalledTimes(1);
  });

  it('失去焦点', () => {
    const onBlur = jest.fn();
    const { getByPlaceholderText } = render(
      <ThemeProvider theme={defaultTheme}>
        <Textarea placeholder="请输入内容" onBlur={onBlur} />
      </ThemeProvider>,
    );

    fireEvent.blur(getByPlaceholderText('请输入内容'));

    expect(onBlur).toBeCalledTimes(1);
  });
});
