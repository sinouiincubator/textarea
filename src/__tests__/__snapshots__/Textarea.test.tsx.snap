// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea 镜像测试 onchange 1`] = `
.c0 {
  min-height: 80px;
  box-sizing: border-box;
  padding: 4px;
  resize: none;
  outline: none;
  font-size: 1rem;
  font-family: "Helvetica Neue",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-weight: 400;
  overflow: auto;
  color: rgba(0,0,0,0.87);
  background-color: transparent;
  border: 1px solid rgba(0,0,0,0.42);
  -webkit-transition: border-color 300ms;
  transition: border-color 300ms;
}

.c0:hover {
  padding: 3px;
  border: 2px solid rgba(0,0,0,0.54);
}

<textarea
  className="c0 sinouiincubator-textarea"
  onBlur={[Function]}
  onChange={[MockFunction]}
  onFocus={[Function]}
  value="001"
/>
`;

exports[`Textarea 镜像测试 渲染Textarea 1`] = `
.c0 {
  min-height: 80px;
  box-sizing: border-box;
  padding: 4px;
  resize: none;
  outline: none;
  font-size: 1rem;
  font-family: "Helvetica Neue",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-weight: 400;
  overflow: auto;
  color: rgba(0,0,0,0.87);
  background-color: transparent;
  border: 1px solid rgba(0,0,0,0.42);
  -webkit-transition: border-color 300ms;
  transition: border-color 300ms;
}

.c0:hover {
  padding: 3px;
  border: 2px solid rgba(0,0,0,0.54);
}

<textarea
  className="c0 sinouiincubator-textarea"
  onBlur={[Function]}
  onFocus={[Function]}
/>
`;
